enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,189 @@
+    * An openshift-krew-index with supported artifacts could hold information about CLIs supported by OpenShift. However, with `krew index add https://someother-third-party-index` 
+    we won't limit cluster-admins from adding their own index with whatever plugins they want.  Only the openshift-krew-index will be supported.
+
+Initial implementation will be with Option 3. Future implementation can combine 3 and 4. Each OpenShift CLI will publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+and also create an image similar to this but that also includes sha256sum files: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+
+Users will continue to have the option to download binaries from either the central location where artifacts are published or from the console.  This proposal adds the mechanism to install from images 
+built from the same artifacts. All options will offer the same binaries, ensured through sha256 signatures.  
+
+For initial implementation, supported CLIs will create ClusterCLI custom resources.  The CLI will be installed to a user's home directory through `oc image extract`.  Through mirrored images and ImageContentSourcePolicy, CLIs will be made available to disconnected environments. 
+
+### Non-Goals
+
+* `oc clustercli` will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* `oc clustercli` will not create or update the ClusterCLI Custom Resources, that will be managed by individual CLIs. 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI manager for various CLIs available for OpenShift and related services so that I can discover, install and list them
+The user will invoke the following commands:
+
+* `oc clustercli list` to view installed binaries
+* `oc clustercli search` will access the connected OpenShift cluster and will retrieve information about available CLIs (clusterCLI CRs)
+* `oc clustercli install odo` will install `odo` to a user's home directory, as a standalone binary.
+
+##### example:
+```
+$ oc clustercli list
+NAME                 DESC                        LATEST       INSTALLED
+-----                -----                       -----        -----
+kubectl              Kubernetes CLI              1.15         1.13
+oc                   OpenShift CLI               4.4          4.3
+odo                  OpenShift Developer CLI     1.0          Not Installed
+kn                   Knative CLI                 1.0          Not Installed
+tkn                  Tekton CLI                  1.0          Not Installed
+
+```
+
```

Comment:
I think we should consider a [Web Terminal](https://www.openshift.com/blog/a-deeper-look-at-the-web-terminal-operator-1) story as the Web Terminal should be a consumer of the OpenShift cluster CLIs.

```
#### Story 2

As a user, I want to access various CLIs from the OpenShift Console Web Terminal so that I can use the
services available in the cluster. The CLIs versions should match the those of the operators deployed 
on the cluster.
```

At the moment Web Terminal CLIs are downloaded at (web terminal) build time. That means that a users will end up using  `oc` v4.5.3 on a 4.6.x cluster. To address that we should be able to consume the mechanism described in this enhancement. Option 3 may work for the Web Terminal operator but this enhancement is lacking the details for consumers that do not want to use `oc` (i.e. some details about the CRD and how a consumer discovers the list of available CLIs and subscribe to updates).

Reply:
I've added this story to the proposal, ptal, thanks!